src/dos/drive_virtual.cpp

```diff
@@ -346,15 +346,14 @@ Virtual_File::Virtual_File(const vfile_data_t& in_data)
 	open = true;
 }
 
-bool Virtual_File::Read(uint8_t * data,uint16_t * size) {
-	uint32_t left = file_data->size() - file_pos;
-	if (left <= *size) {
-		memcpy(data, &((*file_data)[file_pos]), left);
-		*size = (uint16_t)left;
-	} else {
-		memcpy(data, &((*file_data)[file_pos]), *size);
-	}
-	file_pos += *size;
+bool Virtual_File::Read(uint8_t* data, uint16_t* bytes_requested)
+{
+	assert(file_pos <= file_data->size());
+	const uint32_t bytes_left = file_data->size() - file_pos;
+	const uint8_t* src  = file_data->data() + file_pos;
+	*bytes_requested = std::min((uint16_t)bytes_left, *bytes_requested);
```

Comment:
Are these files only 16bit in length?

If not, casting down to 16-bit will wrap bytes left back to 0 at "max-uint16 + 1".

I think we want file sizes to be 64bit (HDD images, being the main use of 64bit sizes).

So maybe store the bytes left in an explicit unint64_t, and also cast up the incoming required bytes up to match.

Reply:
`Virtual_File::Read` gets called by `bool DOS_ReadFile(uint16_t entry,uint8_t * data,uint16_t * amount,bool fcb)` which i think is the implementation for DOS int 21h/AH=3Fh Read file API which does not support larger reads then 16bit values